sdk/compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/Prefixing.hs

```diff
@@ -164,22 +167,19 @@ getProgressLenses = \case
       => ProgressLenses m
     both = ProgressLenses (Just $ Lens $ LSP.params . LSP.workDoneToken) (Just $ Lens $ LSP.params . LSP.partialResultToken)
 
-findKeyByValue :: Eq v => v -> Map.Map k v -> Maybe k
-findKeyByValue val = fmap fst . find ((==val) . snd) . Map.toList
-
--- Remove the prefix from WindowWorkDoneProgressCancel, and return the source SubIDE if we knew it.
+-- Removes and returns the subIDE prefix from cancel messages. Gives Nothing for client created tokens
 removeWorkDoneProgressCancelTokenPrefix
-  :: ProgressTokenPrefixesVar
-  -> LSP.NotificationMessage 'LSP.WindowWorkDoneProgressCancel
-  -> IO (LSP.NotificationMessage 'LSP.WindowWorkDoneProgressCancel, Maybe FilePath)
-removeWorkDoneProgressCancelTokenPrefix tokensVar notif = do
-  let token = notif ^. LSP.params . LSP.token
-  mKey <- findKeyByValue token <$> readMVar tokensVar
-  case mKey of
-    Nothing -> error "Got a cancel request for a ProgressToken that was never created."
-    Just key@(unPrefixedToken, mHome) -> do
-      modifyMVar_ tokensVar (pure . Map.delete key)
-      pure (notif & LSP.params . LSP.token .~ unPrefixedToken, mHome)
+  :: LSP.NotificationMessage 'LSP.WindowWorkDoneProgressCancel
+  -> (LSP.NotificationMessage 'LSP.WindowWorkDoneProgressCancel, Maybe T.Text)
+removeWorkDoneProgressCancelTokenPrefix notif =
+  case removeProgressTokenPrefix $ notif ^. LSP.params . LSP.token of
+    -- Token was created by the client, add the client prefix and broadcast to all subIDEs
+    (unprefixedToken, NoPrefix) ->
+      let prefixedToken = addProgressTokenPrefix ClientPrefix unprefixedToken
+       in (notif & LSP.params . LSP.token .~ prefixedToken, Nothing)
+    -- Created by subIDE, strip the prefix and send to the specific subIDE that created it.
+    (unprefixedToken, SubIDEPrefix prefix) -> (notif & LSP.params . LSP.token .~ unprefixedToken, Just prefix)
+    (_, ClientPrefix) -> error "Client attempted to cancel a ProgressToken with the client prefix, which it should not be aware of. Something went wrong."
```

Comment:
Can we turn this error into a warning somehow? I'd rather that bad cancel requests don't break the IDE too eagerly

Reply:
The problem I have with these cases is that they should be entirely impossible, and if they do occur, something is very wrong with the state of one or more of the subIDEs, and other things will break. If this goes wrong, I want people to tell me, not for it to disappear in a warning log that they ignore.
It also worsens the types, as I'd need to pipe both IO and Maybe down these functions now.
I can still do this, but I wanna make sure you have the same context as I do.
What do you think?

Reply:
Also note that VSCode will restart the multi-IDE if it fully crashes, which I think is right. If something goes this wrong, we want a clean slate.

Reply:
Can we tuck it behind a flag, "ignore-errorneous-subide"? So that users can have an escape hatch if it turns out to be an innocuous issue?

Reply:
Slack discussion: decided against this, as this is a very specific and rare case, and doesn't justify a flag. If this case occurs, a full restart of the multi-ide state is the only "reasonable" course of action, alongside alerting the user (which vscode should)